R/generate_dust.R

```diff
@@ -624,20 +626,49 @@ generate_dust_gpu <- function(eqs, dat, rewrite) {
 
   ## We want to store all time varying things that have rank of at
   ## least 1.
-  i <- vlapply(dat$data$elements, function(x)
-    x$location == "internal" && x$stage == "time" && x$rank > 0)
-  internal <- dat$data$elements[i]
+  internal <- names_if(vlapply(dat$data$elements, function(x)
+    x$location == "internal" && x$stage == "time" && x$rank > 0))
 
   ## It's actually really hard to get integers here, so for now just
   ## assert that there aren't any!
-  stopifnot(vcapply(internal, "[[", "storage_type") == "double")
+  internal_type <- vcapply(dat$data$elements[internal], "[[", "storage_type")
 
-  len <- vcapply(internal, function(x) rewrite(x$dimnames$length))
-  dat$gpu <- list(internal = list(real = len))
+  ## Try and find the things that we actually need in the shared
+  ## object. It won't be all and a bit of efficiency here will go
+  ## a long way.
+  equations <- dat$components$rhs$equations
+  used <- unique(unlist(
+    lapply(dat$equations[equations], function(x) x$depends$variables),
+    FALSE, FALSE))
+
+  ## Also need all offsets and dimensions; could take a subset but
+  ## it's error prone.
+  used_extra <- grep("^(dim_|offset_variable_)", names(dat$data$elements),
+                     value = TRUE)
+  used <- union(used, used_extra)
+
+  pos <- vlapply(dat$data$elements, function(x)
+    x$location == "internal" && x$stage != "time")
+  shared <- intersect(used, names_if(pos))
+  shared_rank <- viapply(dat$data$elements[shared], "[[", "rank")
+  ## Always sort scalars first; this is required later
+  i <- order(shared_rank)
+  shared <- shared[i]
+  shared_rank <- shared_rank[i]
+  shared_type <- vcapply(dat$data$elements[shared], "[[", "storage_type")
+
+  dat$gpu <- list(
+    internal = list(
+      int = intersect(internal, names_if(internal_type == "int")),
+      real = intersect(internal, names_if(internal_type == "double"))),
```

Comment:
They are doubles here, and become real_t later?

Reply:
(looks like it, but just 'double'-checking)

Reply:
yeah, in the underling odin IR we have types "bool", "int", "double" (https://github.com/mrc-ide/odin/blob/master/inst/schema.json#L41-L42) but in dust we reat things that are "double" as "real_t"